olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp

```diff
@@ -155,40 +155,47 @@ ApiClientLookup::ApiClientResponse ApiClientLookup::LookupApi(
   auto client = std::make_shared<client::OlpClient>();
   client->SetSettings(std::move(settings));
 
-  client::Condition condition(cancellation_context);
+  client::Condition condition(
+      std::chrono::seconds(settings.retry_settings.timeout));
 
   // when the network operation took too much time we cancel it and exit
   // execution, to make sure that network callback will not access dangling
   // references we protect them with atomic bool flag.
-  auto interest_flag = std::make_shared<std::atomic_bool>(true);
+  auto flag = std::make_shared<std::atomic_bool>(true);
 
   ApiClientLookup::ApiClientResponse api_response;
+  auto apiClientCallback =
+      [&, flag](ApiClientLookup::ApiClientResponse response) {
+        if (flag->exchange(false)) {
+          api_response = std::move(response);
+          condition.Notify();
+        }
+      };
+
   cancellation_context.ExecuteOrCancelled([&]() {
-    return ApiClientLookup::LookupApiClient(
-        client, service, service_version, catalog,
-        [&,
-         interest_flag](ApiClientLookup::ApiClientResponse response) mutable {
-          if (interest_flag->exchange(false)) {
-            api_response = std::move(response);
-            condition.Notify();
-          }
-        });
+    auto token = ApiClientLookup::LookupApiClient(
+        client, service, service_version, catalog, std::move(apiClientCallback));
+    return client::CancellationToken([&, token]() {
+      if (flag->exchange(false)) {
+        token.cancel();
+        condition.Notify();
+      }
+    });
   });
 
   if (!condition.Wait()) {
```

Comment:
Is it possible in branch to cancel operation twice: once guarded with cancellation context and once unguarded with token.cancel() from :180?

Reply:
the lambda is called from cancellation_context.CancelOperation() itself, so the flag is set during exchange operation and ensures that token is cancelled only once and basically simulates user's call of CancelOperation() after a timeout.